File: ios/ExpoMapboxNavigationModule.swift
import ExpoModulesCore

/// Brand/appearance hooks passed from JS (`theme` prop).
struct NavigationThemeRecord: Record {
  @Field var primary: String?
  @Field var night: Bool?
}

/// Expo module definition: registers the single `ExpoMapboxNavigation` view and
/// its props/events. All real work lives in `MapboxNavigationView`.
public class ExpoMapboxNavigationModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoMapboxNavigation")

    // True when a navigation session is alive (possibly minimised). Reads the
    // nonisolated flag so it stays synchronous.
    Function("isNavigationActive") { () -> Bool in
      NavSessionFlags.active
    }

    // Explicitly end navigation (the trip screen's "Stop" action).
    Function("stopNavigation") {
      Task { @MainActor in NavigationSession.shared.stop() }
    }

    View(MapboxNavigationView.self) {
      Events("onRouteProgress", "onArrival", "onCancel", "onReroute", "onError")

      Prop("coordinates") { (view: MapboxNavigationView, coords: [[Double]]) in
        view.setCoordinates(coords)
      }
      Prop("profile") { (view: MapboxNavigationView, profile: String?) in
        view.setProfile(profile ?? "driving-traffic")
      }
      Prop("mute") { (view: MapboxNavigationView, mute: Bool?) in
        view.setMuted(mute ?? false)
      }
      Prop("theme") { (view: MapboxNavigationView, theme: NavigationThemeRecord?) in
        view.setTheme(theme)
      }
    }
  }
}

File: android/src/main/java/expo/modules/mapboxnavigation/ExpoMapboxNavigationModule.kt
package expo.modules.mapboxnavigation

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo module definition: registers the single `ExpoMapboxNavigation` view and
 * its props/events. All real work lives in [MapboxNavigationView].
 */
class ExpoMapboxNavigationModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoMapboxNavigation")

    // True when a navigation session is alive (possibly minimised).
    Function("isNavigationActive") { NavSession.active }

    // Explicitly end navigation (the trip screen's "Stop" action).
    Function("stopNavigation") { NavSession.stop() }

    View(MapboxNavigationView::class) {
      Events("onRouteProgress", "onArrival", "onCancel", "onReroute", "onError")

      Prop("coordinates") { view: MapboxNavigationView, coords: List<List<Double>> ->
        view.setCoordinates(coords)
      }
      Prop("profile") { view: MapboxNavigationView, profile: String? ->
        view.setProfile(profile ?: "driving-traffic")
      }
      Prop("mute") { view: MapboxNavigationView, mute: Boolean? ->
        view.setMuted(mute ?: false)
      }
      Prop("theme") { view: MapboxNavigationView, theme: Map<String, Any?>? ->
        view.setTheme(theme)
      }
    }
  }
}
